### ISSUE TITLE
Support Unicode variable names

### ISSUE THREAD

**Issue by armgong** [climateinteractive/antlr4-vensim#12]
Currently, it seems antlr4-vensim not support Unicode variable name, but vensim, JavaScript and c++ both support unicode variable name.
Also according https://github.com/antlr/antlr4/blob/master/doc/unicode.md, antlr's  lexer support the full range of Unicode code from 4.7.
So could antlr4-vensim support Unicode variable name? this seems only a slightly modify of the grammar file.

**Comment by chrispcampbell** [comment]
@armgong: Thanks for filing the issue.  This sounds useful.  Any chance you'd be willing to try making the changes to see how extensive the modifications are?  If not, I understand, but it may take me a few weeks before I'll have the time to investigate myself.

One thing to note is that I'm not sure what sort of changes we'll need to make to handle this on the SDEverywhere side, but making the changes on an antlr4-vensim branch first seems like a good way to start.

**Comment by armgong** [comment]
tried  a quick and dirty mod.
test on antlr4-vensim (0.60 because sdeverywhere use this version) and sdeverywhere cli latest version (without change or mod anything) ,convert vensim model to c,  run ok on our model with chinese variable name. 
we only use c/c++ code with vensim model, so not test WebAssembly and webapplication, but it should be ok 

btw those Unicode char code range  copy from other language's lexer, but I really couldn't remember which language.
~~~diff
diff --git a/grammar/Expr.g4 b/grammar/Expr.g4
index 142871b..ecd2384 100644
--- a/grammar/Expr.g4
+++ b/grammar/Expr.g4
@@ -43,11 +43,33 @@ Exclamation : '!' ;
 Id : ( ( Nondigit IdChar* ) | ( Nondigit ( IdChar | ' ' )* IdChar ) | StringLiteral ) ( Whitespace* Exclamation )? ;
 
 fragment
-IdChar : [a-zA-Z0-9_$'&%] ;
+IdChar : Nondigit
+   | '0'..'9'
+   |'_'
+   |'$'
+   |'&'
+   |'\u0027'
+   |'%'
+   |'\u00B7'
+   |'\u0300'..'\u036F'
+   |'\u203F'..'\u2040'
+   ;
 
 fragment
-Nondigit : [a-zA-Z_] ;
-
+Nondigit : 'A'..'Z' | 'a'..'z'
+   |'_'
+   | '\u00C0'..'\u00D6'
+   | '\u00D8'..'\u00F6'
+   | '\u00F8'..'\u02FF'
+   | '\u0370'..'\u037D'
+   | '\u037F'..'\u1FFF'
+   | '\u200C'..'\u200D'
+   | '\u2070'..'\u218F'
+   | '\u2C00'..'\u2FEF'
+   | '\u3001'..'\uD7FF'
+   | '\uF900'..'\uFDCF'
+   | '\uFDF0'..'\uFFFD'
+   ;
 fragment
 Digit
     :   [0-9]
~~~

**Comment by chrispcampbell** [comment]
Hi @armgong, thanks for the suggested fix.  I will run some tests on some of our larger models to make sure there's no significant impact (on performance or otherwise), and if all looks good I'll get this integrated.

**Comment by armgong** [comment]
@chrispcampbell  any update on this ?

**Comment by chrispcampbell** [comment]
Hi @armgong, my apologies, this fell off my list last year and I never completed the work 😞

I picked it back up today and created a new patch similar to the one you proposed, except I based it on the [Swift 5 lexer grammar](https://github.com/antlr/grammars-v4/blob/cc1e848/swift/swift5/Swift5Lexer.g4#L160) from the antlr grammars repo, and cleaned up the naming of the `Id` rules.

I tested the changes with our En-ROADS build and did not see any visible impact on compilation performance.

I filed an issue on the SDE side (https://github.com/climateinteractive/SDEverywhere/issues/532) to pick up the latest version of antlr4-vensim once the fix is published, and I've also added an integration test to verify that the SDEverywhere parse, compile, and runtime packages work fine with non-Latin characters.

I will submit a pull request for this shortly and will get it merged and published after review.

**Comment by armgong** [comment]
thank you very much for the patch, finally no need patch sdeverywhere manually

**Comment by chrispcampbell** [comment]
Hi @armgong, thanks again for your contribution on this.  I merged the fix, published a new antlr4-vensim@0.6.3 that includes the fix, and also published new versions of the following `@sdeverywhere` packages that include the change.

* `@sdeverywhere/parse` v0.1.1
* `@sdeverywhere/compile` v0.7.20
* `@sdeverywhere/cli` v0.7.27
* `@sdeverywhere/create` v0.2.19

Let me know if you run into any issues getting your models to parse after this change.
